File: examples/rust/examples/client_chat_sessions.rs
//! Chat Session Management Example
//!
//! This example demonstrates how to create and manage persistent chat sessions.
//! It shows session creation, messaging, branching, and history management.
//!
//! Prerequisites:
//! - Set OPENAI_API_KEY, ANTHROPIC_API_KEY, or PERPLEXITY_API_KEY in .env
//! - Run the ekoDB server: `make run`
//!
//! Run with: `cargo run --example client_chat_sessions`

use ekodb_client::{
    ChatMessageRequest, Client, CollectionConfig, CreateChatSessionRequest, GetMessagesQuery,
    ListSessionsQuery, Record, UpdateSessionRequest,
};
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Load environment variables
    dotenv::dotenv().ok();

    println!("=== ekoDB Chat Session Management Example ===\n");

    // Create client
    let client = Client::builder()
        .base_url(
            std::env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string()),
        )
        .api_key(std::env::var("API_BASE_KEY")?)
        .build()?;

    let collection = "client_chat_sessions_rust";

    // Step 1: Insert sample data
    println!("=== Inserting Sample Data ===");
    let mut doc = Record::new();
    doc.insert("product", "ekoDB");
    doc.insert(
        "description",
        "A high-performance database product with AI capabilities",
    );
    doc.insert("price", 99);
    client.insert(collection, doc, None).await?;
    println!("✓ Inserted sample product\n");

    // Step 2: Create a chat session
    println!("=== Creating Chat Session ===");
    let session_request = CreateChatSessionRequest::new("openai")
        .model("gpt-4")
        .system_prompt("You are a helpful assistant for product information.")
        .collection(CollectionConfig {
            collection_name: collection.to_string(),
            fields: vec![],
            search_options: None,
        });

    let session = client.create_chat_session(session_request).await?;
    let chat_id = session.chat_id.clone();

    println!("✓ Created session: {}", chat_id);

    // Step 3: Send messages in the session
    println!("=== Sending Messages ===");

    let msg1 = client
        .chat_message(
            &chat_id,
            ChatMessageRequest::new("What products are available?"),
        )
        .await?;
    println!("✓ Message 1 sent");
    println!(
        "  Response: {}\n",
        msg1.responses.first().unwrap_or(&String::new())
    );

    let msg2 = client
        .chat_message(&chat_id, ChatMessageRequest::new("What is the price?"))
        .await?;
    println!("✓ Message 2 sent");
    println!(
        "  Response: {}\n",
        msg2.responses.first().unwrap_or(&String::new())
    );

    // Step 4: Get session messages
    println!("=== Retrieving Session Messages ===");
    let messages_response = client
        .get_chat_session_messages(&chat_id, GetMessagesQuery::new().limit(10).sort("asc"))
        .await?;
    println!("✓ Retrieved {} messages", messages_response.messages.len());
    for (i, msg) in messages_response.messages.iter().enumerate() {
        if let Some(ekodb_client::FieldType::String(role)) = msg.get("role") {
            if let Some(ekodb_client::FieldType::String(content)) = msg.get("content") {
                println!(
                    "  Message {}: [{}] {}",
                    i + 1,
                    role,
                    if content.len() > 60 {
                        format!("{}...", &content[..60])
                    } else {
                        content.to_string()
                    }
                );
            }
        }
    }
    println!();

    // Step 5: Update session metadata
    println!("=== Updating Session ===");
    client
        .update_chat_session(
            &chat_id,
            UpdateSessionRequest::new().system_prompt("You are an expert product consultant."),
        )
        .await?;
    println!("✓ Session updated\n");

    // Step 6: Branch the session (branch at message index 0)
    println!("=== Branching Session ===");
    let branch_request = CreateChatSessionRequest::new("openai")
        .model("gpt-4")
        .branch_from(chat_id.to_string(), 0);

    let branched = client.branch_chat_session(branch_request).await?;
    let branch_id = branched.chat_id.clone();

    println!("✓ Created branch: {}", branch_id);
    println!("  Parent: {}\n", chat_id);

    // Step 7: List all sessions
    println!("=== Listing Sessions ===");
    let sessions = client
        .list_chat_sessions(ListSessionsQuery::new().limit(10).sort("desc"))
        .await?;
    println!("✓ Found {} sessions", sessions.total);
    for (i, sess) in sessions.sessions.iter().take(3).enumerate() {
        println!(
            "  Session {}: {} ({})",
            i + 1,
            sess.chat_id,
            sess.title.as_ref().unwrap_or(&"Untitled".to_string())
        );
    }
    println!();

    // Step 8: Get specific session
    println!("=== Getting Session Details ===");
    let session_details = client.get_chat_session(&chat_id).await?;
    println!("✓ Session details retrieved");
    println!("  Messages: {}\n", session_details.message_count);

    // Step 9: Delete the branched session
    println!("=== Deleting Branch Session ===");
    client.delete_chat_session(&branch_id).await?;
    println!("✓ Deleted branch session: {}\n", branch_id);

    // Cleanup: Delete the collection (chat sessions are managed by server)
    println!("=== Cleanup ===");
    client.delete_collection(collection).await?;
    println!("✓ Deleted collection\n");

    println!("✓ All session management operations completed successfully");

    Ok(())
}

File: examples/kotlin/examples/ClientChatSessions.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.types.Record
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject

/**
 * Chat Sessions example - Session management and history
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_chat_sessions_example"
    
    println("=== ekoDB Kotlin Client - Chat Session Management Example ===\n")
    
    try {
        // Step 1: Insert sample data
        println("=== Inserting Sample Data ===")
        
        val doc = Record.new()
            .insert("product", "ekoDB")
            .insert("description", "A high-performance database product with AI capabilities")
            .insert("price", 99)
        client.insert(collection, doc)
        println("✓ Inserted sample product\n")
        
        // Step 2: Create a chat session
        println("=== Creating Chat Session ===")
        
        val sessionRequest = buildJsonObject {
            putJsonArray("collections") {
                add(buildJsonObject {
                    put("collection_name", collection)
                    putJsonArray("fields") {}
                })
            }
            put("llm_provider", "openai")
            put("llm_model", "gpt-4")
            put("system_prompt", "You are a helpful assistant for product information.")
        }
        
        val session = client.createChatSession(sessionRequest)
        val chatId = session["chat_id"]?.jsonPrimitive?.content ?: error("No chat_id in response")
        println("✓ Created session: $chatId\n")
        
        // Step 3: Send messages
        println("=== Sending Messages ===")
        
        val msg1Request = buildJsonObject {
            put("message", "What products are available?")
        }
        val msg1 = client.chatMessage(chatId, msg1Request)
        println("✓ Message 1 sent")
        println("  Responses: ${msg1["responses"]}\n")
        
        val msg2Request = buildJsonObject {
            put("message", "What is the price?")
        }
        val msg2 = client.chatMessage(chatId, msg2Request)
        println("✓ Message 2 sent")
        println("  Responses: ${msg2["responses"]}\n")
        
        // Step 4: Get message history
        println("=== Getting Message History ===")
        
        val messages = client.getChatSessionMessages(chatId)
        println("✓ Retrieved message history")
        println("  Total messages: ${messages["messages"]}\n")
        
        // Step 5: Update session
        println("=== Updating Session ===")
        
        val updateRequest = buildJsonObject {
            put("system_prompt", "You are an expert product consultant.")
        }
        client.updateChatSession(chatId, updateRequest)
        println("✓ Updated session system prompt\n")
        
        // Step 6: List all sessions
        println("=== Listing Sessions ===")
        
        val sessions = client.listChatSessions()
        println("✓ Total sessions: ${sessions["total"]}\n")
        
        // Step 7: Branch session
        println("=== Branching Session ===")
        
        val branchRequest = buildJsonObject {
            putJsonArray("collections") {
                add(buildJsonObject {
                    put("collection_name", collection)
                    putJsonArray("fields") {}
                })
            }
            put("llm_provider", "openai")
            put("llm_model", "gpt-4")
            put("parent_id", chatId)
        }
        
        val branchedSession = client.branchChatSession(branchRequest)
        val branchedChatId = branchedSession["chat_id"]?.jsonPrimitive?.content ?: error("No chat_id in response")
        println("✓ Created branched session: $branchedChatId\n")
        
        // Cleanup
        println("=== Cleanup ===")
        client.deleteChatSession(chatId)
        client.deleteChatSession(branchedChatId)
        println("✓ Deleted chat sessions")
        
    } finally {
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n✓ Chat session management example completed successfully")
    }
}
